inference/fim-inference.py

```diff
@@ -0,0 +1,130 @@
+    name = p.name.replace(".json", ".results.json")
+    return Path(name)
+
+def dump_json_gz(p: Path, obj):
+    """
+    dumps an object to json.
+    """
+    if p.suffix == ".gz":
+        with gzip.open(p, "wt") as f:
+            f.write(json.dumps(obj, indent=4))
+    else:
+        with open(p, "w") as f:
+            f.write(json.dumps(obj, indent=4))
+
+def generate_completions(prompt_file: Path, args, model):
+    prompts = read_json_gz(prompt_file)
+    processed_prompts = []
+    for prompt in tqdm(prompts, unit="files"):
+        results = []
+        body = prompt["body"]
+        header = prompt["prompt"]
+        for line_index in tqdm(range(len(body)), unit="lines"):
+            line = body[line_index]
+            prefix = header + "\n".join(body[:line_index])
+            suffix = "\n".join(body[line_index + 1:])
+            submitted_prompt = f"{FIM_PREFIX}{prefix}{FIM_SUFFIX}{suffix}{FIM_MIDDLE}"
+            new_completions = model.completions(
+                prompt=submitted_prompt,
+                max_tokens=args.max_tokens,
+                temperature=args.temperature,
+                n=args.batch_size,
+                top_p=args.top_p,
+                stop=["\n"],
+            )
+            results.append({ 
+                "line": line,
+                "results": [{"completion": completion, "correctness": completion.strip() == line.strip()} 
+                                for completion in new_completions]
+                }
+            )
+        
+        processed_prompts.append({
+            "language": prompt["language"],
+            "prompt": prompt["prompt"],
+            "body": prompt["body"],
+            "results": results
+        })
+
+    return processed_prompts
+
+def main():
+
+    args = argparse.ArgumentParser()
+
+    args.add_argument(
+        "--input-files",
+        type=Path,
+        nargs="*",
+        help="files used for input.",
+    )
+
+    args.add_argument(
+        "--output-dir",
+        type=Path,
+        help="Directory in which to place JSON files with completions. The default is fim-humaneval",
+        default="fim-humaneval"
+    )
+
+    args.add_argument(
+        "--max-tokens", type=int, default=100, help="Maximum number of tokens to generate"
+    )
+    
+    args.add_argument(
+        "--temperature", type=float, default=0.8, help="Temperature of completions"
+    )
+
+    args.add_argument(
+        "--batch-size", type=int, default=16, help="Number of completions to batch"
+    )
+
+    args.add_argument(
+        "--top-p", type=float, default=0.95, help="Number of completions to batch"
+    )
+
+    args.add_argument(
+        "--model-name",
+        type=str,
+        required=True,
+        help="The model name. To add a new model, copy and modify codegen.py",
+    )
+
+    args = args.parse_args()
+
+    model = importlib.import_module(args.model_name)
+
+    model.model.add_special_tokens(SPEC_TOKS)
```

Comment:
I think we should edit all the bigcode_models (except NOFIM and MHA) to just add these as special tokens. So, maybe the constructor for `Model` should just have a `supports_fims` flag.